lib/dao.js

```diff
@@ -624,7 +622,208 @@ DataAccessObject.upsert = function(data, options, cb) {
   }
   return cb.promise;
 };
+/**
+ * Performs upsert based on the instance that matches the where clause
+ . This is a PUT operation and based on
+ * non-primary key. The filter takes one key value pair and updates single matching instance.*
+ */
+DataAccessObject.patchOrCreateWithWhere =
+DataAccessObject.upsertWithWhere = function(where, data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) { return connectionPromise; }
+  assert(arguments.length >= 1, 'At least one argument is required');
+  if (data === undefined && options === undefined && cb === undefined && arguments.length === 1) {
+    data = where;
+    where = {};
+  } else if (options === undefined && cb === undefined) {
+        // One of:
+        // updateAll(data, cb)
+        // updateAll(where, data) -> Promise
+    if (typeof data === 'function') {
+      cb = data;
+      data = where;
+      where = {};
+    }
+  } else if (cb === undefined) {
+        // One of:
+        // updateAll(where, data, options) -> Promise
+        // updateAll(where, data, cb)
+    if (typeof options === 'function') {
+      cb = options;
+      options = {};
+    }
+  }
+  data = data || {};
+  options = options || {};
+  cb = cb || utils.createPromiseCallback();
+  assert(typeof where === 'object', 'The where argument must be an object');
+  assert(typeof data === 'object', 'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+  var hookState = {};
+  var filter = { 'where': where };
+  var self = this;
+  var Model = this;
+  var connector = Model.getConnector();
+  var modelObject = connector._models;
+  var modelName = Object.keys(modelObject)[0];
+  var id = getIdValue(this, data);
+  var context = {
+    Model: Model,
+    query: { 'where': where },
+    hookState: hookState,
+    options: options,
+  };
+  Model.notifyObserversOf('access', context, doUpsertWithNPK);
+  function doUpsertWithNPK(err, ctx) {
+    if (err) return cb(err);
+    var isOriginalQuery = isWhereByGivenId(Model, ctx.query.where, id);
```

Comment:
`isOriginalQuery` check is needed only when the original intent was to update a single instance matching a given id, to ensure that a hook did not changed the query to match more than one instance, or a different instance.

In the case of `upsertWithWhere`, I think we can allow hooks to change the `where` condition to match different set of models, because the code below already handles all different outcomes (no model found, one model found, multiple models found).